app/lib/entity_cache.rb

```diff
@@ -19,22 +19,28 @@ def emoji(shortcodes, domain)
     shortcodes = Array(shortcodes)
     return [] if shortcodes.empty?
 
-    cached       = Rails.cache.read_multi(*shortcodes.map { |shortcode| to_key(:emoji, shortcode, domain) })
+    domain = domain.downcase if domain
+
+    cached       = Rails.cache.read_multi(*shortcodes.map { |shortcode| to_emoji(:emoji, shortcode, domain) })
     uncached_ids = []
 
     shortcodes.each do |shortcode|
-      uncached_ids << shortcode unless cached.key?(to_key(:emoji, shortcode, domain))
+      uncached_ids << shortcode unless cached.key?(to_emoji(:emoji, shortcode, domain))
     end
 
     unless uncached_ids.empty?
       uncached = CustomEmoji.where(shortcode: shortcodes, domain: domain, disabled: false).index_by(&:shortcode)
-      uncached.each_value { |item| Rails.cache.write(to_key(:emoji, item.shortcode, domain), item, expires_in: MAX_EXPIRATION) }
+      uncached.each_value { |item| Rails.cache.write(to_emoji(:emoji, item.shortcode, domain), item, expires_in: MAX_EXPIRATION) }
     end
 
-    shortcodes.filter_map { |shortcode| cached[to_key(:emoji, shortcode, domain)] || uncached[shortcode] }
+    shortcodes.filter_map { |shortcode| cached[to_emoji(:emoji, shortcode, domain)] || uncached[shortcode] }
   end
 
   def to_key(type, *ids)
     "#{type}:#{ids.compact.map(&:downcase).join(':')}"
   end
+
+  def to_emoji(type, *ids)
+    "#{type}:#{ids.compact.join(':')}"
```

Comment:
I'm not sure what's happening differently here vs the existing method, which already downcases the incoming keys/ids? The domain gets downcased there for rails key cache purposes, and elsewhere for sql query/storage purposes...

Reply:
When the pr was written, mastodon regards two emojis with different cases but same spelling as one same emoji (in to_key method) which will cause it to render a single post with two different emojis unexpectedly. So to_emoji won't downcase them (though domain names should be downcased which I'll fix it later). Anyhow, I'm not sure if this bug has been fixed in recent releases, after all, it's almost two years later. My instance currently has not two emojis with same spelling and two cases, so I'll check if this (potentially erroneous behavior still exists)